### PR TITLE
Bring recent stk development into Trilinos.

### PR DESCRIPTION
**Pull request by alanw0**
@trilinos/<teamName>

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@trilinos/stk

**Comment by mhoemmen** [comment]
Hm:
```
CMake Error at cmake/tribits/core/package_arch/TribitsAdjustPackageEnables.cmake:143 (MESSAGE):
  Error, the package 'STKNGP_TEST' is listed as a dependency of the package
  'STKUnit_tests' is in the list 'Trilinos_SE_PACKAGES' but the package
  'STKNGP_TEST' is either not defined or is listed later in the package
  order.  This may also be an attempt to create a cicular dependency between
  the packages 'STKNGP_TEST' and 'STKUnit_tests' (which is not allowed).
  Check the spelling of 'STKNGP_TEST' or see how it is listed in
  Trilinos_PACKAGES_AND_DIRS_AND_CLASSIFICATIONS in relationship to
  'STKUnit_tests'.
```

**Comment by lucbv** [comment]
@alanw0 it seems that the last failure was legitimate so I'm putting the `AT: WIP` label.
The auto-tester is really slow at this point so there is no need to have it retest this before a fix is performed...

**Comment by alanw0** [comment]
@mhoemmen or @prwolfe please review once again...

File: packages/stk/stk_ngp_test/stk_ngp_test/Reporter.hpp [hunk 1]
```diff
@@ -13,13 +13,13 @@ namespace ngp_testing {
 class TruncatedString {
  public:
   static constexpr int maxNumChar = 63;
-  NGP_TEST_FUNCTION TruncatedString() : string("") {}
+  NGP_TEST_FUNCTION TruncatedString() {}
```
Comment (mhoemmen) [inline packages/stk/stk_ngp_test/stk_ngp_test/Reporter.hpp]: `NGP_TEST_FUNCTION TruncatedString() = default;`

**Comment by alanw0** [comment]
Most of the tests passed, that's progress. Looks like there's a remaining error for cuda, which somehow doesn't show up for sierra's nightly cuda builds as far as I know. I'll activate this again once I figure that out.

**Comment by mhoemmen** [comment]
@alanw0 wrote:
> Looks like there's a remaining error for cuda, which somehow doesn't show up for sierra's nightly cuda builds as far as I know.

@theguruat12 and @rrdrake might want to know if the CUDA build in the PR tests doesn't match your CUDA build.

**Comment by jwillenbring** [comment]
@alanw0 @mhoemmen

> Looks like there's a remaining error for cuda, which somehow doesn't show up for sierra's nightly cuda builds as far as I know.

Please report back when you figure out what is different because we should consider whether changes are necessary. In general when setting up the cuda build we had trouble with STK tests, so we disabled them all. We also disabled seacas.

**Comment by alanw0** [comment]
@mhoemmen or @prwolfe I made a change to a source file to try to solve the undefined extern error. If that doesn't work I'll have to dig deeper into the mysterious cuda build.

**Comment by prwolfe** [comment]
Still seeing this today. Not sure if I will have time to work on it today.

**Comment by alanw0** [comment]
@prwolfe The fix I made yesterday was to pull a small constructor definition from the .cpp file to the .hpp. That fixed the build failure for the constructor in question. But last night another constructor caused the same error. So I'm guessing the cuda build isn't using the --rdc flag. Up until now all of stk's cuda code has been header only. But the new stk_ngp_test package has some code in .cpp files. So I think we need to either add --rdc (relocatable device code) to Trilinos' cuda build, or we'll have to make stk_ngp_test be header only. I prefer the --rdc option currently. (assuming that is indeed the issue here)

**Comment by alanw0** [comment]
@jwillenbring Following up from this morning's sierra dev-ops meeting. I think this pull-request might go through successfully if the --rdc option were provided to the cuda build. Who is in charge of the settings for the PR builds? I see some files under cmake/std which look like the place to make changes, but it's not clear to me how to change build flags in there, nor if I'm even qualified or allowed to do that.

**Comment by alanw0** [comment]
@prwolfe or @mhoemmen I made some changes to the new stk_ngp_test library to make it usable as header-only. I don't know if a new thumbs-up from you is required to kick off the PR testing again.

**Comment by alanw0** [comment]
@prwolfe or @mhoemmen It finally built! Quick, approve it before something changes!

**Comment by mhoemmen** [comment]
@alanw0 I'm in meetings all week, but not now :D

**Comment by prwolfe** [comment]
thank you @mhoemmen

**Pull request merged**